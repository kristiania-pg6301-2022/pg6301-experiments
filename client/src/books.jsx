import * as React from "react";
import {useEffect, useState} from "react";
import {Route, Routes} from "react-router-dom";
import {useNavigate} from "react-router";

export function Books() {
    const [books, setBooks] = useState([]);

    async function load() {
        const res = await fetch("/api/books");
        setBooks(await res.json());
    }

    useEffect(async () => {
        load();
    })


    async function handleAddBook(book) {
        const response = await fetch("/api/books", {
            "method": "POST",
            body: JSON.stringify(book),
            headers: {
                "content-type": "application/json"
            }
        });
        console.log(response.status);
        await load();
    }


    return <>
        <Routes>
            <Route path={"/"} element={<ListBooks books={books}/>}/>
            <Route path={"create"} element={<CreateNewBook onAddBook={handleAddBook}/>}/>
        </Routes>
    </>;
}

export function CreateNewBook({onAddBook}) {
    const navigate = useNavigate();

    async function handleSubmit(e) {
        e.preventDefault();
        await onAddBook({author, title, year});
        navigate("..");
    }

    const [author, setAuthor] = useState("");
    const [title, setTitle] = useState("");
    const [year, setYear] = useState("");

    return <div>
        <h1>Add book to server</h1>
        <form onSubmit={handleSubmit}>
            <div>
                <label>Author: <input data-testid="author" value={author} onChange={e => setAuthor(e.target.value)}/></label>
            </div>
            <div>
                <label>Title: <input data-testid="title" value={title} onChange={e => setTitle(e.target.value)}/></label>
            </div>
            <div>
                <label>Year: <input data-testid="year" value={year} onChange={e => setYear(e.target.value)}/></label>
            </div>
            <button>Submit</button>
        </form>
    </div>;
}

export function ListBooks({books}) {
    return <div>
        <h1>{books.length} books</h1>
        {books.map(({author, id, title, year}) => <div key={id}>{author}: {title} ({year})</div>)}
    </div>;
}

