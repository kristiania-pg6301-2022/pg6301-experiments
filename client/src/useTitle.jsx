import {useEffect} from "react";

export function useTitle(title, excludeApplicationName = false) {
    const originalTitle = document.title;
    useEffect(() => {
        let newTitle = title;
        if (!excludeApplicationName) {
            newTitle += " | Kristiania Library";
        }
        document.title = newTitle;
        return () => {
            document.title = originalTitle;
        }
    }, []);
}
