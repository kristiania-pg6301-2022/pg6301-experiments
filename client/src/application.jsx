import * as React from "react";
import {BrowserRouter, Link, Route, Routes} from "react-router-dom";
import {Books} from "./books";
import {useTitle} from "./useTitle";

function ApplicationHeader() {
    return <header><h1><Link to={"/"}>Kristiania Library</Link></h1></header>;
}

function ApplicationMenu() {
    return <nav>
        <ul>
            <li><Link to={"books"}>View books</Link></li>
            <li>View authors</li>
            <li><Link to={"books/create"}>Add book</Link></li>
        </ul>
    </nav>;
}

function FrontPage() {
    useTitle("Kristiania Library", true);
    return <h1>Welcome</h1>;
}

function ApplicationMain() {
    return <main>
        <div>
            <Routes>
                <Route path={"/"} element={<FrontPage/>}/>
                <Route path={"books/*"} element={<Books/>}/>
                <Route path={"*"} element={<h1>Not found</h1>}/>
            </Routes>
        </div>
    </main>;
}

export function Application() {
    return <BrowserRouter>
        <main>
            <ApplicationHeader/>
            <ApplicationMenu/>
            <ApplicationMain/>
        </main>
    </BrowserRouter>
}

