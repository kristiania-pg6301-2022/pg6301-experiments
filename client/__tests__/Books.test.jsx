import * as React from "react";
import * as ReactDOM from "react-dom";
import {act} from "react-dom/test-utils";

import {Books, CreateNewBook, ListBooks} from "../src/books";
import {MemoryRouter} from "react-router";

describe("Books", () => {

    it("shows create book form", async () => {
        const container = document.createElement("div");
        await act(async () => {
            ReactDOM.render(
                <MemoryRouter><CreateNewBook/></MemoryRouter>,
                container
            )
        })
        expect(container.innerHTML).toMatchSnapshot();
    });

    it("shows book list", async () => {
        const books = [
            {id: 1, author: "Johannes", title: "React lectures", year: 2021},
            {id: 2, author: "Johannes", title: "Java lectures", year: 2021},
        ];
        const container = document.createElement("div");
        await act(async () => {
            ReactDOM.render(<ListBooks books={books}/>, container)
        })
        expect(container.innerHTML).toMatchSnapshot();
    });

});

