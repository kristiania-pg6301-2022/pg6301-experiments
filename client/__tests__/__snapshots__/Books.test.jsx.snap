// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books shows book list 1`] = `"<div><h1>2 books</h1><div>Johannes: React lectures (2021)</div><div>Johannes: Java lectures (2021)</div></div>"`;

exports[`Books shows create book form 1`] = `"<div><h1>Add book to server</h1><form><div><label>Author: <input data-testid=\\"author\\" value=\\"\\"></label></div><div><label>Title: <input data-testid=\\"title\\" value=\\"\\"></label></div><div><label>Year: <input data-testid=\\"year\\" value=\\"\\"></label></div><button>Submit</button></form></div>"`;
